rule/rule.go

```diff
@@ -126,19 +183,31 @@ func ScanAST(pkg string, bzlFile *bzl.File) *File {
 		Pkg:  pkg,
 		Path: bzlFile.Path,
 	}
-	for i, stmt := range f.File.Stmt {
-		switch stmt := stmt.(type) {
+	f.Rules, f.Loads, _ = ScanASTBody("", bzlFile.Stmt)
+	f.Directives = ParseDirectives(bzlFile)
+	return f
+}
+
+// ScanASTBody scans a list of Starlark expressions and returns the rules,
+// loads and if a defName is supplied it will try to return a function
+// matching that name. This data can be wrapped by a File and later modified.
+func ScanASTBody(defName string, stmt []bzl.Expr) (rules []*Rule, loads []*Load, fn *bzl.Function) {
```

Comment:
I think the signature should be `ScanASTBody(pkg, defName string, bzlFile *bzl.File) *File`.

If `defName == ""`, it should do what `ScanAST` did previously. `ScanAST` can just return `ScanASTBody(pkg, "", bzlFile)`.

If `defName != ""`, then it should take care of the stuff `LoadMacroData` is doing now. `LoadMacroData` can just parse the file, then call this function.